### PR TITLE
feat: Support running tests from Java Project explorer

### PR DESCRIPTION
**Pull request by jdneo**
Support run tests from the Java Project explorer

### Demo

https://user-images.githubusercontent.com/6193897/104272416-05afab80-54d8-11eb-9542-03ff98789c5b.mp4

### Note
Since we need to run all the children nodes if the explorer is in Hierarchical mode, so I made some workaround in the implementation. Basically it's to patch the arguments resolved by the configuration delegator.

### CODE REVIEW

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java [hunk 1]
```diff
@@ -354,36 +359,57 @@ private static void searchInFolder(List<TestItem> resultList, SearchTestItemPara
...
+
+    /**
+     * Get all sub-packages whose name starts with the given package.
+     */
+    public static IPackageFragment[] getAllSubPackages(IJavaElement packageFragment) throws CoreException {
+        final List<IPackageFragment> result = new ArrayList<>();
+        if (packageFragment != null && packageFragment instanceof IPackageFragment) {
+            final IJavaElement[] packages = ((IPackageFragmentRoot) packageFragment.getParent()).getChildren();
+            for (final IJavaElement p : packages) {
+                if (!(p instanceof IPackageFragment)) {
```
Comment (Eskibear) [inline java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:403]: looks better in style?
```
if (p instanceof IPackageFragment) 
   if (p.getElementName().startsWith(packageFragment.getElementName()))
            result.add((IPackageFragment) p);
```

**Review by Eskibear — DISMISSED** [review]

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java [hunk 2]
```diff
@@ -354,36 +359,57 @@ private static void searchInFolder(List<TestItem> resultList, SearchTestItemPara
...
         }
 
-        return null;
+        return element;
+    }
+
+    /**
+     * Get all sub-packages whose name starts with the given package.
+     */
+    public static IPackageFragment[] getAllSubPackages(IJavaElement packageFragment) throws CoreException {
```
Comment (testforstephen) [inline java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:398]: this method actually returns the package itself and its children, so a better name should be `getPackageAndSubpackages`.

Comment (testforstephen) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:398]: There is duplicated utility in jdt.core.manipulation jar.

See org.eclipse.jdt.internal.corext.refactoring.util.JavaElementUtil.getPackageAndSubpackages(), suggest to reuse it.

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java (hunk 1 above)
Comment (jdneo) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:403]: Got the point, or maybe merge the if blocks into one?

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java (hunk 2 above)
Comment (jdneo) [reply java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java:398]: Yes that will be much better.

**Review by testforstephen — APPROVED** [review]

**Pull request merged**